### DISCUSSION TITLE
Execution Limits as a Function of Bitcoin Block Times

### DISCUSSION THREAD

**Discussion by gregorycoppola** [stacks-network/stacks-core#3241] [Ideas]
## Abstract
We propose that **optimal throughput** on 1) the Stacks chain, and 2) proprietary subnets, depends on the ability for the **execution limit budget in a given Bitcoin block round** to be a **function** of the **length of time** that round (which is not under Stacks' control) takes.

We analyze some data about Bitcoin block times to motivate this statement.

## Bitcoin Block Times
Bitcoin is timed so that Blocks are produced *on average* every 10 minutes. However, the variance and distribution of the actual Bitcoin block times is not shaped like a normal distribution, but like an [exponential distribution](https://blog.lopp.net/bitcoin-block-time-variance/):

![exponential-distribution](https://user-images.githubusercontent.com/60008382/183128426-1b905e05-8a09-4643-a82b-5abfb5e77f77.jpeg)

This means that we have:
* lots of blocks that will be smaller than the average
* some blocks that will be very long

Empirically, this is what we find in this scrape from Dec 1, 2021:
![image](https://user-images.githubusercontent.com/60008382/183128929-b783239e-5ced-41d0-aa68-f52990b6747a.png)

While a only *small minority of blocks* take more than 10 minutes (38%), a *large majority* (75%) of the *time spent* is in blocks that take longer than 10 minutes. During these "long blocks" latency is bad, and throughput is impaired.

## Execution Limits
Execution limits are currently determined based on the smallest blocks. There has to be enough time to 1) process the previous block, and 2) make a new block, between two Bitcoin blocks.

The limit is currently targeted to take around 30 seconds of processing. This means, in the 75% of *time* that is spent in blocks longer than 10 minutes, we only get 30 seconds of processing done. This includes in intervals that could be 10, 20 or even 30 minutes.

Fixed execution cost limits that are appropriate for short blocks (which are common) will lose time in long blocks.

## Variable Execution Cost Limits
**Variable execution cost limits** would mean that the execution budget for the Stacks chain can grow dynamically during a long block. Somehow, as the Bitcoin block gets longer, the execution cost limit should be increased.

This post asks:
* is it possible to have **variable execution cost limits**?:
  * on the L1 Stacks chain, this would have to happen in a trustless way
    * e.g., multiple miners take turns during a Bitcoin epoch
  * on the L2 Hiro subchain, this could happen in a trusted way
    * e.g., federation members are trusted to say T time has passed, so the limit can be increased

**Comment by jcnelson** [comment]
> Somehow, as the Bitcoin block gets longer, the execution cost limit should be increased.

It isn't possible to determine in an absolute sense how long a Bitcoin block took to mine.

First, there is know way to predict _a priori_ how long it will take until the next block is discovered.  We have a probability function, but that does not tell us anything about particular blocks.  So, any such Bitcoin block time determination protocol will require participating nodes to somehow agree on how much wall-clock time was spent waiting for a block once it arrives.  The need to get nodes to agree on when a Bitcoin block arrived creates problems, which I'll elaborate on below.

Second, such a protocol cannot exist in an open-membership setting and do better than Bitcoin by more than a constant factor, because there can be no global consensus on the _true time_ at which the block arrives, and thus there can be no global consensus on _how long it took_.  There are a couple reasons for this:

* Bitcoin miners' clocks are not tightly-synchronized.  In fact, the only constraints on a newly-arrived Bitcoin block's reported timestamp are that it (1) is greater than the median timestamp of the previous 11 blocks and that it (2) is within 2 hours if the node's network-adjusted time, which is a function of its local clock and its neighbors' clocks (see https://en.bitcoin.it/wiki/Block_timestamp).  Because Bitcoin is open-membership -- i.e. there's no constraints on who can be a node's neighbor -- there are no protocol-enforced limits on the delta between a local node's time and the true time.  The delta between when your Bitcoin node reports the arrival of a block and when mine does can be arbitrarily big.

* Valid Bitcoin blocks can arrive at a node at arbitrary times in the future after they were mined (e.g. someone can mine a valid fork privately, and release it later), which precludes the creation of an agreement protocol as to what time a block was produced.  There's no way to know at what time the private fork's miner actually produced the block, since no one besides the miner witnessed the event.

What this means is that while it may be possible for Stacks miners watching Bitcoin blocks arrive to execute some sort of voting protocol to decide how long it took for the Bitcoin block to propagate (e.g. to within one RTT), there's no way to authenticate the decisions after the fact.  What this in turn means is that whoever are the miners _today_ could vote to decide that a Bitcoin block took 1 day to arrive (or 1 week, or 1 month, etc.), and in doing so, allot themselves arbitrarily many epochs between two Bitcoin blocks and therefore prevent the rest of the network from processing the next Bitcoin block's sortitions (and in doing so, prevent _future_ miners from joining).  Therefore, such a block time decision protocol would transform Stacks into a closed-membership system, which we do not want.

While it is possible for Stacks nodes to participate in a Bitcoin block time determination protocol built on Nakamoto consensus, such a protocol would _de facto_ re-implement Bitcoin.  Nodes would gain no more certainty about the arrival times of Bitcoin blocks from this protocol, because they would in turn be unable to agree on the arrival times of their own messages in this protocol to a degree of precision that is superior to Bitcoin's by any more than a constant factor.

I think if you want variable block execution times, you're going to have to come up with a way to do so that doesn't depend on watching the Bitcoin chain.

**Reply by gregorycoppola** [reply]
It's definitely a non-trivial problem.
